ext/http/00_serve.js

```diff
@@ -331,25 +331,76 @@ function fastSyncResponseOrStream(req, respBody) {
 }
 
 async function asyncResponse(responseBodies, req, status, stream) {
-  const responseRid = core.ops.op_set_response_body_stream(req);
-  SetPrototypeAdd(responseBodies, responseRid);
   const reader = stream.getReader();
-  core.ops.op_set_promise_complete(req, status);
+  let hasResolved = false;
   try {
-    while (true) {
-      const { value, done } = await reader.read();
-      if (done) {
-        break;
+    // Optimize for streams that are done in zero or one packets. We will not
+    // have to allocate a resource in this case.
+
+    // We read the first value as normal
+    const { value: value1, done: done1 } = await reader.read();
+    if (done1) {
+      core.ops.op_set_promise_complete(req, status);
+      return;
+    }
+
+    // The second value cannot block indefinitely, as someone may be waiting on a response
+    // of the first packet that may influence this packet. We set this timeout arbitrarily to 250ms
+    // and we race it.
+    let read2 = reader.read();
+    let timeoutHandle;
+    const timeout = new Promise((r) => {
+      timeoutHandle = setTimeout(r, 250);
+    });
+    const winner = Promise.race([read2, timeout]);
```

Comment:
Aren't you missing an await here?

Reply:
Newer version of the code changes this race.